src/templates/next/layouts/Content/Content.tsx

```diff
@@ -1,33 +1,140 @@
+  permalink: string[],
+): BreadcrumbProps => {
+  const breadcrumb = []
+  let node = sitemap
+  let currentPath = ""
+  for (const pathSegment of permalink) {
+    currentPath += "/" + pathSegment
+    const nextNode = node.children?.find(
+      (node) => node.permalink === currentPath,
+    )
+    if (!nextNode) {
+      // TODO: handle this unexpected case where cannot traverse to permalink in the sitemap
+      break
+    }
+    node = nextNode
+    breadcrumb.push({
+      title: node.title,
+      url: node.permalink,
+    })
+  }
+  return { links: breadcrumb }
+}
+
+const getSiderailFromSiteMap = (
+  sitemap: IsomerSitemap,
+  permalink: string[],
+): SiderailProps | null => {
+  let node = sitemap
+  let currentPath = ""
+
+  let i = 0
+  while (i < permalink.length - 1) {
+    currentPath += "/" + permalink[i]
+    const nextNode = node.children?.find(
+      (node) => node.permalink === currentPath,
+    )
+    if (!nextNode) {
+      // TODO: handle this unexpected case where cannot traverse to parent in the sitemap
+      return null
+    }
+    node = nextNode
+    i++
+  }
+  if (!node.children) {
+    // TODO: handle this unexpected case where parent does not contain current page
+    return null
+  }
+  const parentTitle = node.title
+  const parentUrl = node.permalink
+
+  const pages = []
+  // get all siblings of page
+  const pagePath = "/" + permalink.join("/")
+  for (const sibling of node.children) {
+    if (sibling.permalink === pagePath) {
+      pages.push({
+        title: sibling.title,
+        url: sibling.permalink,
+        isCurrent: true,
+        childPages: sibling.children?.map((child) => ({
+          url: child.permalink,
+          title: child.title,
+        })),
+      })
+    } else {
+      pages.push({
+        title: sibling.title,
+        url: sibling.permalink,
+      })
+    }
+  }
+  return {
+    parentTitle,
+    parentUrl,
+    pages,
+  }
+}
+
+const getTableOfContentsFromContent = ({
+  content,
+}: Pick<ContentPageSchema, "content">) => {
+  const items = []
+  for (const block of content) {
+    if (block.type === "heading" && block.level === 2) {
+      items.push({
+        content: block.content,
+        anchorLink: "#" + block.id,
+      })
+    }
+  }
+  return { items }
+}
 
 const ContentLayout = ({
   site,
   page,
   content,
   LinkComponent,
 }: ContentPageSchema) => {
+  const sideRail = getSiderailFromSiteMap(
+    site.siteMap,
+    page.permalink.split("/").slice(1),
+  )
+  const tableOfContents = getTableOfContentsFromContent({ content })
+  const breadcrumb = getBreadcrumbFromSiteMap(
+    site.siteMap,
+    page.permalink.split("/").slice(1),
+  )
   return (
-    <Skeleton site={site} page={page}>
-      <div className="lg:hidden">
-        {page.sideRail && <Siderail {...page.sideRail} />}
-      </div>
+    <Skeleton site={site} page={page} LinkComponent={LinkComponent}>
+      <div className="lg:hidden">{sideRail && <Siderail {...sideRail} />}</div>
```

Comment:
nit: I think the `siteRail` control flow should be outside the `<div>` tag?

Reply:
[de186df](https://github.com/isomerpages/isomer-components/pull/62/commits/de186df46d9ffcad61d6bc410c02601ace8e6d0c)